tutorials/haskell/mustache-templates/tutorial.md

```diff
@@ -0,0 +1,404 @@
+
+## Rendering
+
+There is just one function that you need to render your template —
+`renderMustache`. Its type signature looks like this:
+
+```haskell
+renderMustache :: Template -> Value -> Text
+```
+
+First you feed it with a `Template` (we just discovered how to get these),
+then you need a source of values to interpolate. For that we need a `Value`
+which comes from Aeson — a library for working with JSON in Haskell. Why to
+do it this way? Well, Mustache templates originate from languages with duck
+typing, so to feed values into a Mustache template, we need to categorize
+them to understand how they should interact with the template we are
+rendering. Aeson's `Value` is exactly a type that is close to what you have
+in Ruby (and of course in JavaScript), and since Aeson is a very popular
+package, we already know how to convert pretty much any type to `Value`.
+Isn't that cool?
+
+To interpolate something we need to have it first. Let's put together some
+data:
+
+```haskell
+release :: Release
+release = Release
+  { reGenre  = "Ambient"
+  , reDate   = "1980"
+  , reDiscId = "380B7905"
+  , reComment = "ExactAudioCopy v0.95b4"
+  , rePerformer = "Laraaji"
+  , reTitle = "Ambient 3 Day Of Radiance"
+  , reFiles =
+    [ File
+      { fiFileName = "01 - The Dance #1.wav"
+      , fiIndex = 1
+      , fiTitle = "The Dance #1"
+      , fiPerformer = "Laraaji"
+      , fiIndex00 = "00:00:00" }
+    , File
+      { fiFileName = "02 - The Dance #2.wav"
+      , fiIndex = 2
+      , fiTitle = "The Dance #2"
+      , fiPerformer = "Laraaji"
+      , fiIndex00 = "09:06:10" }
+    ]
+  }
+```
+
+Just use the `renderMustache` function now with `toJSON` (which transforms
+instances of `ToJSON` to `Value`):
+
+```
+main :: IO ()
+main = do
+  let template = $(TH.compileMustacheDir "main" "data/")
+  TIO.putStrLn $ renderMustache template (toJSON release)
+```
+
+As simple as that, if you run the program, you will see that our CUE file
+looks good:
+
+```
+REM GENRE "Ambient"
+REM DATE "1980"
+REM DISCID "380B7905"
+REM COMMENT "ExactAudioCopy v0.95b4"
+PERFORMER "Laraaji"
+TITLE "Ambient 3 Day Of Radiance"
+FILE "01 - The Dance #1.wav" WAVE
+  TRACK 1 AUDIO
+    TITLE "The Dance #1"
+    PERFORMER "Laraaji"
+    INDEX 00 00:00:00
+FILE "02 - The Dance #2.wav" WAVE
+  TRACK 2 AUDIO
+    TITLE "The Dance #2"
+    PERFORMER "Laraaji"
+    INDEX 00 09:06:10
+```
+
+How does the magic work? What to do if you want to construct `Value`
+manually? Simply put, `toJSON` does not do anything you can't do yourself.
+Records are transformed into JSON `Object`s, lists and vectors and turned
+into `Array`s, etc. Feeding the `renderMustache` without `toJSON` is just as
+easy, consult
+[Aeson documentation](https://hackage.haskell.org/package/aeson) for more
+information, but the most common thing you will want to do is to create a
+custom dictionary that will be your “context”, it's done with help of
+`object` and `(.=)` functions:
```

Comment:
@mrkkrp 

- "if you want to construct a `Value`..."
- "(consult Aeson's documentation for more information)" or "-- consult Aeson's documentation for more information --"
- "your context, which it's done with help of the..." or "your context. This is done with help of the..."